\documentclass[12pt]{article}
\usepackage{geometry}                
\geometry{letterpaper}                   % ... or a4paper or a5paper or ... 
\usepackage[parfill]{parskip}    
\usepackage{graphicx}
\usepackage{amssymb}
\usepackage{amsmath}
\usepackage[usenames,dvipsnames,svgnames,table]{xcolor}


\title{Sequencing Uncertainty Propagation}
\author{David Champredon}
\date{\today}                                           % Activate to display a given date or no date

\newcommand{\comment}[1]{\textsl{\textcolor{cyan}{#1}}}
\newcommand{\sq}[1]{\texttt{\textcolor{brown}{#1}}}
\newcommand{\pr}[1]{\mathrm{Pr}(#1)}


\begin{document}
\maketitle

\section{Background}

Identifying the sequence of nucleotides from a biological sample is a complex process which is fraught with noise. 

Assuming the biological sample of interest has been properly isolated (that is it is complete, has no damages or contamination), sequencing a biological sample, whether with the Sanger method or ``Next-Generation Sequencing'' (NGS) usually involves:
\begin{itemize}
\item if the sample of interest is RNA: Reverse transcription 
\item DNA fragmentation in smaller pieces than the original sample (less than 400bp for NGS, 800bp for Sanger)
\item amplification of the fragmented DNA using PCR
\item sequencing the fragments (identifying the nucleotides from a fluorescent tag attached)
\item alignment or mapping: putting back the small fragment together by aligning them (de novo) or mapping them on benchmark libraries
\end{itemize}


Errors can be introduced at each of these steps for various reasons \cite{Beerenwinkel:2011}. It is probably not feasible to determine what is the source of the noise, nor to try to eliminate it completely.
The goal here is to acknowledge there is uncertainty in the output sequence given from any sequencing method and to propose a method to propagate this uncertainty in any downstream analysis.
Currently, this uncertainty is recognized and even quantified with sequencing quality scores (FASTQ files), but it does not seem those scores are used to inform a probabilistic model to represent the sequence. 
Simply put, we shouldn't treat the result of sequencing as a \emph{certainty}.




\section{Probabilistic representation}

\subsection{Definition}
 
We can represent probabilistically a nucleotide sequence in a matrix form. For a sequence of length $\ell$ we can write:

$$S = \bordermatrix{   & 1 & 2 & \ldots & \ell \cr
                \sq{A} & p_{A, 1} & p_{A, 2} & \ldots & p_{A, \ell} \cr
                \sq{C} & p_{C, 1} & p_{C, 2} & \ldots & p_{C, \ell} \cr
                \sq{G} & p_{G, 1} & p_{G, 2} & \ldots & p_{G, \ell} \cr
                \sq{T} & p_{T, 1} & p_{T, 2} & \ldots & p_{T, \ell} \cr 
                \sq{x} & p_{x, 1} & p_{x, 2} & \ldots & p_{x, \ell} \cr 
}$$

Each column represents the nucleotide position, each row one of the four nucleotide \sq{A,C,G,T} as well as an empty position \sq{x}.
Hence, $S$ is a $5\times\ell$ matrix. Its elements represent the probability that a nucleotide is at given position:

\begin{equation}
S_{n,j} = \pr{\text{nucleotide \sq{n} is at position }j}
\end{equation}
with the special case for a deletion:

\begin{equation}
S_{\sq{x},j} = \pr{\text{empty position }j}
\end{equation}

The matrix $S$ will be called the \emph{probabilistic sequence} of a biological sample.
Note, that we have, for all $1\leq j \leq \ell$:
\begin{equation}
\sum_{n\in \{ \sq{A,C,G,T,x} \} } S_{n, j} = 1
\end{equation}

The sequence of nucleotides from a biological sample is not treated as certainty anymore, but as a collection of possible sequences, with length not necessarily equal when the probability of an empty position is positive. 



\subsection{Examples}

If we have the following probabilistic sequence
$$
S = 
\begin{pmatrix}
0.9 & 0.05   & 0.99 & 0 & 0\\
0   & 0.8 & 0 & 0 & 0.1\\
0.1 & 0.15 & 0 & 0.3 & 0.9\\
0 & 0 & 0.01 & 0.7 & 0\\
0 & 0 & 0 & 0 & 0\\
\end{pmatrix}
$$
then there are $2\times 3 \times 2^3 = 48$ possible sequences. The most likely is the one having the highest nucleotides probabilities: \sq{ACATG} with probability 0.449  ($0.9\times 0.8\times 0.99 \times 0.7 \times 0.9$).

If there is a positive probability for at least one empty position, then the sequence has a variable length. 
Let's take the same example as above, but adding one possible empty position:

$$
S = 
\begin{pmatrix}
0.9 & 0.05   & 0.99 & 0 & 0\\
0   & 0.8 & 0 & 0 & 0.1\\
0.1 & 0.15 & 0 & 0.2 & 0.9\\
0 & 0 & 0.01 & 0.7 & 0\\
0 & 0 & 0 & 0.1 & 0\\
\end{pmatrix}
$$

Like above, there is still a 0.449 probability that the sequence is \sq{ACATG}, but with probability 0.064, the sequence could be shorter when position 4 is empty and be \sq{ACAG}.



\section{Quantifying probabilities}

The difficulty of estimating the probabilities that populate the probabilistic sequence lies in quantifying errors at each steps (fragmentation, amplification, sequencing, alignment). And how to integrate these error types in a coherent fashion?


\subsection{Sequencing errors}

Maybe the easiest error type to quantify is the fragment sequencing error because a quality (or ``Phred'') score is attributed to each base call. The quality score $Q$ is directly related to the error probability: $p = 10^{-Q/10}$. 

\begin{figure}[ht]
% https://tex.stackexchange.com/questions/511138/miktex-graphics-version-1-3b-bug-on-windows-10-setcurrfile-undefined

%\begin{center}
% \includegraphics{figs/Phred.jpg}
%\caption{default}
%\label{default}
%\end{center}
\end{figure}

As I don't have FASTQ files from actual biological samples, I use the software  \textsf{inSilicoSeq} that simulates the sequencing of short reads from Illumina instruments. 
As shown in Figure \comment{TODO: include figure, but fix Miktex graphics version first}, the sequencing error probability ranges between $10^{-3.5}$ an $10^{-1.5}$ .  

After alignment, taking an optimistic view, we can assume a unique global sequencing error of $\epsilon=10^{-3.5}=0.0003$.
So each base call is right with probability $1-\epsilon$. Assume the other bases and missing position \sq{x} are all equally likely with probability $(1-\epsilon)/4$. 

For example, if the output sequence after fragment sequencing and alignment is \sq{ACGT} the probabilistic sequence is:

$$
S = 
\begin{pmatrix}
1-\epsilon & \epsilon/4    & \epsilon/4  & \epsilon/4  \\
\epsilon/4 & 1-\epsilon & \epsilon/4  & \epsilon/4  \\
\epsilon/4 & \epsilon/4  & 1-\epsilon  & \epsilon/4 \\
\epsilon/4 & \epsilon/4  & \epsilon/4  & 1-\epsilon \\
\epsilon/4 & \epsilon/4  & \epsilon/4  & \epsilon/4 \\
\end{pmatrix}
$$





\end{document}  